tests/ctst/HOW_TO_WRITE_TESTS.md

```diff
@@ -0,0 +1,121 @@
+or renaming files. You must ensure the scenarios are fully idempotent.
+"Flakies" are mostly due to non-idempotent tests, then bugs. It is rarely due
+to the platform.
+
+**Rule #2**
+
+> Use unique resources names. S3 Bucket, Account and S3 Objects should be
+> unique.
+
+It limits the risks of collision between tests.
+For cross-account scenarios, use unique S3 Policies and S3 Roles.
+
+If data is used, such as lock files, ensure they are named appropriately.
+
+**Rule #3**
+
+> Avoid reconfiguring the environment during test execution.
+
+If a reconfiguration triggers a reconciliation of the resources, running tests
+might be affected. Also, it restarts the services, which leads to missing logs,
+harder debugging, and longer test execution.
+
+Examples:
+
+- Location creation;
+- Bucket Website endpoints.
+
+... must be created before the test start, or in a dedicated set of test.
+
+**Rule #4**
+
+> Do not use `atMostOnePicklePerTag`.
+
+If a set of scenario requires the use of `atMostOnePicklePerTag`, they won't
+be executed in parallel, which is:
+
+- Not realistic for the production environment.
+- Not efficient for the test execution: the duration of tests will suffer.
+- A proof that the test is not following **Rule #1**.
+
+Ensuring this rule may be hard, and idempotent scenarios are not always
+possible. Solutions exist:
+
+- Pre-create the resources required for the tests before they start.
+- Use `Before` / `After` hooks to setup your environment once.
+  Concurrency can be controlled with lock files.
+- Make your assertions more flexible: strict and absolute checks are good
+  for unit or functional testing, but in integration tests, prefer using
+  relative checks.
+- As a last resort, we might have a dedicated test suite.
+
+**Rule #5**
+
+> Avoid scenarios with no value.
+
+It is tempting to write scenarios that test more than what is needed
+at the integration testing layer: please trust your unit and functional
+tests. If a scenario has too many variants, then likely they can be tested
+in lower layers. Please focus on validating a feature.
+
+**Rule #6**
+
+> Ensure assertions are idempotent after sleep times.
+
+Some tests might need to "sleep" for some time, either static or using
+more complex logic. The subsequent assertions must take into account
+any operation on the platform that might be asynchronous.
+
+Example: we run a Kubernetes job in tests that look idempotent. However,
+the job itself might not be idempotent and run in parallel, leading to
+some resources being skipped by a previous job that would take longer than
+expected, and the test results will differ from what is expected.
+
+**Rule #7**
+
+> Avoid using the @Flaky tag.
+
+If a test is marked as flaky, it is likely not idempotent, or we did not
+do our job correctly to understand the problem. Please fix the test.
+
+If an issue exists in the product, prefer fixing the bug first.
```

Comment:
```suggestion
Whenever there is a flaky, there is an issue : either a bug (corner case or race condition) in the product,
or simply an issue in the test. So the issue should be fixed either way.

More often than not, retrying the build (i.e. ignoring the flaky) or automating the retry with @Flaky is just
masking a bug, and leaving it in production instead of fixing it : so it must not be done.

One frequent problem in flaky tests is that they are not idempotent, or that we did not do our job correctly to understand the problem. Flaky tests must be investigated, and issues in the tests must be fixed to ensure we
don't ignore an actual problem in the product.
```